remme/clients/basic.py

```diff
@@ -194,31 +194,41 @@ def _socket_request(self, suffix, data=None):
             raise ClientException('Suffix "%s" not supported' % suffix)
 
     def _handle_response(self, msg_type, resp_proto, req):
-        self._stream.wait_for_ready()
-        future = self._stream.send(
-            message_type=msg_type,
-            content=req.SerializeToString())
+        # Loops while NOT_READY
+        while True:
+            self._stream.wait_for_ready()
 
-        resp = resp_proto()
-        try:
-            resp.ParseFromString(future.result().content)
-        except (DecodeError, AttributeError):
-            raise ClientException(
-                'Failed to parse "content" string from validator')
-        except ValidatorConnectionError as vce:
-            LOGGER.error('Error: %s' % vce)
-            raise ClientException(
-                'Failed with ZMQ interaction: {0}'.format(vce))
-
-        data = message_to_dict(resp)
+            future = self._stream.send(
+                message_type=msg_type,
+                content=req.SerializeToString())
 
-        # NOTE: Not all protos have this status
-        with suppress(AttributeError):
-            if resp.status == resp_proto.NO_RESOURCE:
-                raise KeyNotFound("404")
+            resp = resp_proto()
 
-        if resp.status != resp_proto.OK:
-            raise ClientException("Error: %s" % data)
+            try:
+                resp.ParseFromString(future.result().content)
+            except (DecodeError, AttributeError):
+                raise ClientException(
+                    'Failed to parse "content" string from validator')
+            except ValidatorConnectionError as vce:
+                LOGGER.error('Error: %s' % vce)
+                raise ClientException(
+                    'Failed with ZMQ interaction: {0}'.format(vce))
+
+            data = message_to_dict(resp)
+
+            # NOTE: Not all protos have this status
+            with suppress(AttributeError):
+                if resp.status == resp_proto.NO_RESOURCE:
+                    raise KeyNotFound("404")
+
+            if resp.status != resp_proto.OK:
+                LOGGER.info(f'_handle_response: error: {data}')
+                if hasattr(resp_proto, 'NOT_READY') and resp_proto.NOT_READY == resp.status:
+                    time.sleep(5)
```

Comment:
Really??

Reply:
Periodically attempts to _handle_response. It is well tested

Reply:
Did u test with concurrent connections? It is the same question with blocking the event loop. Use the async version of `time`. And read please about what is the event loop and how it works.

Reply:
I don't think our rest api is asynchronous, it currently relies on multithreading

Reply:
No, ws, http server already on asyncio.

Reply:
Anyway it is a bad practice to do such things in single thread

Reply:
@bogomazov REST API is based on asyncio, so please be asynchronous.

Reply:
@eugene-babichenko if it was synchronous, `sleep` is bad anyway to use in this situation.